Layout floats as children of their inline ancestors

mrobinson opened the pull request:
When layout was split into two phases, floats were laid out as direct
children of the inline formatting context. This meant that they were
positioned properly, but not properly made children of their inline
ancestors' stacking contexts. This change maintains the proper
positioning of floats, but positions them relatively to their inline
ancestors.

The big change here is that `text-align` needs to be taken into account
before actually laying out LineItems. This has the added benefit of
setting inline layout for the implementation of `text-align: full`. Now
all line items are laid out at the real final position and we can adjust
the `start_corner` property of float `BoxFragments` when their ancestors
are laid out.

<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] There are tests for these changes

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

mrobinson committed:
Layout floats as children of their inline ancestors

When layout was split into two phases, floats were laid out as direct
children of the inline formatting context. This meant that they were
positioned properly, but not properly made children of their inline
ancestors' stacking contexts. This change maintains the proper
positioning of floats, but positions them relatively to their inline
ancestors.

The big change here is that `text-align` needs to be taken into account
before actually laying out LineItems. This has the added benefit of
setting inline layout for the implementation of `text-align: full`. Now
all line items are laid out at the real final position and we can adjust
the `start_corner` property of float `BoxFragments` when their ancestors
are laid out.

mukilan commented on components/layout_2020/flow/inline.rs:
```diff
@@ -174,29 +174,78 @@ impl<'box_tree, 'a, 'b> InlineFormattingContextState<'box_tree, 'a, 'b> {
+        }
+    }
+
+    /// Given the amount of whitespace trimmed from the start of the line and taking into consideration
```
The comment here says 'amount of whitespace trimmed from the *start of the line*' but in code, the value for whitespace_trimmed is set by the 'trim_whitespace_at_end` method. Is the comment here wrong or am I misunderstanding it?

mrobinson replied:
Oh, that's a good point. I've updated this to smiply say "the amount of whitespace trimmed from the line."

mukilan commented on components/layout_2020/flow/inline.rs:
```diff
@@ -235,48 +284,21 @@ impl<'box_tree, 'a, 'b> InlineFormattingContextState<'box_tree, 'a, 'b> {
+        match text_align {
+            TextAlign::Start => line_start_position_from_indent,
+            TextAlign::End => {
+                containing_block.inline_size - real_line_length_without_indent +
```
Is the calculation here correct?
For [this test html](http://software.hixie.ch/utilities/js/live-dom-viewer/saved/11941), both master and legacy-layout have the same output as Firefox and Chrome, but this PR doesn't match those.

mrobinson replied:
The calculation was incorrect! I have updated this in a way that I think corrects the case you mentioned and also have put a specification link and quote explaining why it's correct now.

mukilan replied:
The `text-align: end` case  doesn't seem to work similar to the other layouts & master for [this case](http://software.hixie.ch/utilities/js/live-dom-viewer/?saved=11943).

IIUC, the spec says 'text-indent' is treated as margin at the 'start'  edge of the linebox and 'text-align: end' will align the contents to the 'end' edge of the box, so the margin cannot not influence the end edge unless the margin exceeds the available width.

mrobinson commented:
@bors-servo try=wpt-2020

mrobinson commented:
@mukilan Thanks for the review. I think this is ready to look at again, when you have a moment.

mukilan commented:
Thanks @mrobinson for updating the PR with the spec and comments! I've found one more case where 'text-end' doesn't work as other browsers. Please let me know if that is not in scope for this PR and needs to fixed separately as part of 'text-indent' stuff.

mrobinson commented:
> Thanks @mrobinson for updating the PR with the spec and comments! I've found one more case where 'text-end' doesn't work as other browsers. Please let me know if that is not in scope for this PR and needs to fixed separately as part of 'text-indent' stuff.

Thanks for pointing out this issue. I believe I've fixed it and have also added two WPT tests corresponding to the issues your raised above.

servo-wpt-sync commented:
🤖 Opened new upstream WPT pull request (web-platform-tests/wpt#41585) with upstreamable changes.

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#41585).

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#41585).

mrobinson commented:
It seems that Layout 2013 does not pass this test...it looks like due to line breaking issues. All pipelines are passing in upstream WPT issue. I've updated the result and will send it back to the MQ.

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#41585).
